pkg/kubelet/volumemanager/reconciler/reconstruct_new_test.go

```diff
@@ -0,0 +1,261 @@
+
+	tests := []struct {
+		name                                string
+		volumePaths                         []string
+		postSyncStatCallback                func(rcInstance *reconciler, fakePlugin *volumetesting.FakeVolumePlugin) error
+		expectedVolumesNeedReportedInUse    []string
+		expectedVolumesNeedDevicePath       []string
+		expectedVolumesFailedReconstruction []string
+		verifyFunc                          func(rcInstance *reconciler, fakePlugin *volumetesting.FakeVolumePlugin) error
+	}{
+		{
+			name: "when two pods are using same volume and both are deleted",
+			volumePaths: []string{
+				path.Join("pod1", "volumes", "fake-plugin", "pvc-abcdef"),
+				path.Join("pod2", "volumes", "fake-plugin", "pvc-abcdef"),
+			},
+			expectedVolumesNeedReportedInUse:    []string{"fake-plugin/pvc-abcdef", "fake-plugin/pvc-abcdef"},
+			expectedVolumesNeedDevicePath:       []string{"fake-plugin/pvc-abcdef", "fake-plugin/pvc-abcdef"},
+			expectedVolumesFailedReconstruction: []string{},
+			verifyFunc: func(rcInstance *reconciler, fakePlugin *volumetesting.FakeVolumePlugin) error {
+				mountedPods := rcInstance.actualStateOfWorld.GetMountedVolumes()
+				if len(mountedPods) != 0 {
+					return fmt.Errorf("expected 0 certain pods in asw got %d", len(mountedPods))
+				}
+				allPods := rcInstance.actualStateOfWorld.GetAllMountedVolumes()
+				if len(allPods) != 2 {
+					return fmt.Errorf("expected 2 uncertain pods in asw got %d", len(allPods))
+				}
+				volumes := rcInstance.actualStateOfWorld.GetPossiblyMountedVolumesForPod("pod1")
+				if len(volumes) != 1 {
+					return fmt.Errorf("expected 1 uncertain volume in asw got %d", len(volumes))
+				}
+				return nil
+			},
+		},
+		{
+			name: "when reconstruction fails for a volume, volumes should be cleaned up",
+			volumePaths: []string{
+				path.Join("pod1", "volumes", "missing-plugin", "pvc-abcdef"),
+			},
+			expectedVolumesNeedReportedInUse:    []string{},
+			expectedVolumesNeedDevicePath:       []string{},
+			expectedVolumesFailedReconstruction: []string{"pvc-abcdef"},
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			tmpKubeletDir, err := os.MkdirTemp("", "")
+			if err != nil {
+				t.Fatalf("can't make a temp directory for kubeletPods: %v", err)
+			}
+			defer os.RemoveAll(tmpKubeletDir)
+
+			// create kubelet pod directory
+			tmpKubeletPodDir := filepath.Join(tmpKubeletDir, "pods")
+			os.MkdirAll(tmpKubeletPodDir, 0755)
+
+			mountPaths := []string{}
+
+			// create pod and volume directories so as reconciler can find them.
+			for _, volumePath := range tc.volumePaths {
+				vp := filepath.Join(tmpKubeletPodDir, volumePath)
+				mountPaths = append(mountPaths, vp)
+				os.MkdirAll(vp, 0755)
+			}
+
+			rc, fakePlugin := getReconciler(tmpKubeletDir, t, mountPaths)
+			rcInstance, _ := rc.(*reconciler)
+
+			// Act
+			rcInstance.reconstructVolumes()
+			if tc.postSyncStatCallback != nil {
+				err := tc.postSyncStatCallback(rcInstance, fakePlugin)
```

Comment:
These tests don't use this callback. May as well remove them?

Reply:
Fixed